virttest/qemu_devices/qcontainer.py

```diff
@@ -3091,3 +3093,82 @@ def _emulator_setup(binary, extra_options=None):
         devices.append(tpm_model_dev)
 
         return devices
+
+    def numa_hmat_lb_define_by_params(self, nodeid, initiator, params):
+        """
+        Create numa hmat-lb device by params
+
+        :param nodeid: The target node id of the hmat-lb device
+        :param initiator: The initiator of the hmat-lb device
+        :param params: The params for hmat-lb
+        :return: The qdevices.QCustomDevice of hmat-lb
+        """
+        aobject = '%s_hmat_lb' % nodeid
+        hmat_lb_params = {'target': nodeid, 'initiator': initiator,
+                          'hierarchy': params['numa_hmat_lb_hierarchy'],
+                          'hmat_type': 'hmat-lb'}
+        data_type = params['numa_hmat_lb_data_type']
+        hmat_lb_params.update({'data-type': data_type})
+        if 'latency' in data_type:
+            hmat_lb_params.update(
+                {'latency': params['numa_hmat_lb_latency']})
+            aobject += '_latency'
+        elif 'bandwidth' in data_type:
+            hmat_lb_params.update(
+                {'bandwidth': params['numa_hmat_lb_bandwidth']})
+            aobject += '_bandwidth'
+        return qdevices.QCustomDevice('numa', params=hmat_lb_params,
+                                      aobject=aobject, backend='hmat_type')
+
+    def numa_hmat_cache_define_by_params(self, nodeid, params):
+        """
+        Create numa hmat-cache device by params
+
+        :param nodeid: The target node id of the hmat-cache device
+        :param params: The numa params for hmat-caches
+        :return: The list of qdevices.QCustomDevice, including all hmat-caches
+        targeted on the nodeid given
+        """
+        numa_hmat_caches = params.objects('numa_hmat_caches')
+        if numa_hmat_caches:
+            if (not self.get_by_properties(
+                    {'aobject': '%s_hmat_lb_bandwidth' % nodeid}) or
+                    not self.get_by_properties(
+                        {'aobject': '%s_hmat_lb_bandwidth' % nodeid})):
+                raise exceptions.TestError(
+                    'Please make sure both hmat-lb bandwidth and '
+                    'hmat-lb latency are defined when define hmat-cache.')
+
+        hmat_caches = {}
+        hmat_cache_devs = []
+        aobject = '%s_hmat_cache' % nodeid
+        op = gt if params.get_boolean('numa_hmat_cache_increase') else lt
```

Comment:
I thought this part is handled in case cfg or py? Framework just need to deal with the given params.

Reply:
@huangyum Okay I can remove it, and just write correct definition for `hmat-cache`s.

Reply:
@huangyum Updated together with https://github.com/autotest/tp-qemu/pull/2543 and test passed, please help re-review, thanks.

Reply:
Forgot to push?

Reply:
@huangyum Yes...sorry, just pushed.